src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -274,6 +295,10 @@ def make_requirement_from_install_req(self, ireq, requested_extras):
             name=canonicalize_name(ireq.name) if ireq.name else None,
             version=None,
         )
+        if cand is None:
+            if not ireq.name:
```

Comment:
Maybe a comment here: "Unnamed URL - we can't satisfy this any other way, so fail eagerly"